### PR TITLE
feat(mojaloop/#4000): timeout handler

### PR DESCRIPTION
**Pull request by oderayi**

### CODE REVIEW

File: src/lib/proxyCache.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+const getConnectedCache = async () => {
+  await connect()
+  return proxyCache
+}
+
+const getClient = async () => {
+  return (await getConnectedCache()).client
+}
+
+module.exports = {
```
Comment (geka-evk) [inline src/lib/proxyCache.js:65]: Why do we need this wrapper on top of _proxyCache_?

File: src/handlers/TimeoutHandler.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+let isRunning
+
+const timeout = async () => {
+  if (isRunning) return
+  try {
+    isRunning = true
+    Logger.isDebugEnabled && Logger.debug('Timeout handler triggered')
+    await TimeoutService.timeoutInterschemePartiesLookups()
+  } catch (err) {
+    Logger.isErrorEnabled && Logger.error(err)
```
Comment (geka-evk) [inline src/handlers/TimeoutHandler.js:50]: Can we stop using such approach for logging errors?  it's really hard to debug it. So in my opinion, it should be somethig like:
```
Logger.isErrorEnabled && Logger.error(`error in timeoutInterschemePartiesLookups: ${err?.stack})
```
As far as I remember, we stared the discussion about error logging: https://infitx.slack.com/archives/C04HG5L8LDQ/p1721305796588129

File: src/domain/timeout/index.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+} = require('@mojaloop/event-sdk')
+const Config = require('../../lib/config')
+
+const timeoutInterschemePartiesLookups = async () => {
+  const match = 'als:*:*:*:expiresAt' // als key expiry pattern
+  const count = Config.HANDLERS_TIMEOUT_BATCH_SIZE
+  const redis = await ProxyCache.getClient()
+
+  return Promise.all(
+    redis.nodes('master').map(async (node) => {
```
[the remark this answers is not in the document]
Comment (oderayi) [reply src/domain/timeout/index.js:57]: You're right @vijayg10. I realized that late. I moved the logic to the lib.

File: src/api/index.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+
+ * ModusBox
+ - Rajiv Mothilal <rajiv.mothilal@modusbox.com>
```
[the remark this answers is not in the document]
Comment (oderayi) [reply src/api/index.js:21]: Not really, the file was only moved to a new location.

File: src/handlers/TimeoutHandler.js (hunk 2 above)
Comment (geka-evk) [reply src/handlers/TimeoutHandler.js:50]: BTW, in ALS we already can use improved logger:
```
const { logger } = const Config = require('../lib')
...
logger.error('error in timeoutInterschemePartiesLookups:', err)
```
here's how it looks in stdout output:

File: src/lib/proxyCache.js (hunk 1 above)
Comment (oderayi) [reply src/lib/proxyCache.js:65]: Because it requires initialization (configuration and connection) before it can be used.

Comment (geka-evk) [reply src/lib/proxyCache.js:61]: but we do the same without wrapper:
https://github.com/mojaloop/account-lookup-service/pull/484/files#diff-79f4c7cdc0b026f6e556196f5911a449157807cffdb3f9d33f0b5404b4ffc2f3R98

Seems, you tend to use singletone pattern (with hardcoded dependencies) rather then dependency injection

File: src/handlers/TimeoutHandler.js (hunk 2 above)
Comment (oderayi) [reply src/handlers/TimeoutHandler.js]: Updated.

File: src/domain/timeout/index.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    finishSpan(span, fspiopError)
+    throw fspiopError
+  }
+}
+
+const validateParticipant = async (fspId) => {
+  const participant = await Participant.validateParticipant(fspId)
+  if (!participant) {
+    const errMessage = ERROR_MESSAGES.partyDestinationFspNotFound
+    Logger.isErrorEnabled && Logger.error(`error in validateParticipant: ${errMessage?.stack}`)
```
Comment (geka-evk) [inline src/domain/timeout/index.js:84]: `errMessage` is just a sting, maybe like this:
```
Logger.isErrorEnabled && Logger.error(errMessage)
```

File: src/handlers/register.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+const registerHandlers = async (handlers) => {
+  await init()
+  handlers.forEach(handler => {
+    switch (handler) {
+      case HANDLER_TYPES.TIMEOUT:
+        Logger.isDebugEnabled && Logger.debug('Registering Timeout Handler')
+        TimeoutHandler.register()
+        break
+      default:
+        Logger.isDebugEnabled && Logger.debug(`Handler ${handler} not found`)
```
Comment (geka-evk) [inline src/handlers/register.js:51]: maybe, it's better to use `warn` logLevel for such information?

File: test/integration/domain/timeout/index.test.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    const redisNodes = redisClient.nodes ? redisClient.nodes('master') : [redisClient]
+    await Promise.all(redisNodes.map(async node => {
+      await node.flushall()
+    }))
+  })
+
+  afterAll(async () => {
+    return Promise.all([
+      proxyClient.deleteHistory(),
+      proxyCache.disconnect()
```
Comment (geka-evk) [inline test/integration/domain/timeout/index.test.js:24]: One of disadvantages of using _ProxyCache_ singleton is that we can't now run integration tests in parallel, coz one test might finish and disconnect cache, while another test is trying to read or write some data (and will fail)

File: src/handlers/TimeoutHandler.js (hunk 2 above)
Comment (oderayi) [reply src/handlers/TimeoutHandler.js]: Updated

File: test/integration/domain/timeout/index.test.js (hunk 7 above)
Comment (oderayi) [reply test/integration/domain/timeout/index.test.js:26]: Refactored.

I don't think we can run integration tests in parallel if the datastore and mock-server history is shared among tests.

File: src/handlers/TimeoutHandler.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+  const { logger } = options
+
+  try {
+    isRunning = true
+    logger.debug('Timeout handler triggered')
+    await TimeoutService.timeoutInterschemePartiesLookups(options)
+  } catch (err) {
+    logger.error('error in timeout: ', err)
+    throw ErrorHandler.Factory.reformatFSPIOPError(err)
```
Comment (geka-evk) [inline src/handlers/TimeoutHandler.js:53]: How and where are we going to handle this error?

**Review by vijayg10 — APPROVED** [review]

**Pull request merged**